### ISSUE TITLE
Does not parse template strings within functions within template strings

### ISSUE THREAD

**Issue by psalv** [typescript-eslint/typescript-eslint#2736]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

Added a failing test case here: https://github.com/psalv/typescript-eslint/commit/1f5d20967f72eaec0bfb0eb9c16a8eb58e051e37

I would expect this code block to be parsable:
```TS
const Foo = ` ${() => `test`} `;
```

**Expected Result**

Code block can be parsed without error.

Especially given the following code block, where the only change is making the template string into a regular string, parses fine:
```TS
const Foo = ` ${() => 'test'} `;
```

**Actual Result**

There is an assertion error:
` AssertionError [ERR_ASSERTION]: A fatal parsing error occurred: Parsing error: ',' expected. `

**Additional Info**

This is especially relevant for linting styled-components where the following would be a common code block:
```TS
const StyledElement = styled.div<{display: string}>`
   ${props => `display: ${props.display}`};
`;
```

Note: this only happens when we have a template string within a function within a template string. The following code block parses fine:
```TS
const Foo = ` ${`test`} `;
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `4.6.1` |
| `TypeScript`                | `4.0.2` |
| `ESLint`                    | `7.7.0` |
| `node`                      | `12.18.3` |

**Comment by bradzacher** [comment]
your test code is wrong.

```diff
  it('within arrow function nested in another literal', () => {
    // Does not throw
-   expect(parseAndAnalyze(`const Foo = \` ${() => 'test'} \`;`)).toBeDefined();
+   expect(parseAndAnalyze(`const Foo = \` \${() => 'test'} \`;`)).toBeDefined();

    // Throws
-   expect(parseAndAnalyze(`const Foo = \` ${() => `test`} \`;`)).toBeDefined();
+   expect(parseAndAnalyze(`const Foo = \` \${() => \`test\`} \`;`)).toBeDefined();
  });
```

`` `const Foo = \` ${() => `test`} \`;` `` does not create the template string you'd expect.
It does not create the string ``"const Foo = ` ${() => `test`} `;"``
It instead creates the string ``"const Foo = ` () => `test` `;"``

Hence the parser error because `` ` () => `test` ` `` is not valid JS.

**Comment by psalv** [comment]
Ah! That makes sense.

Thanks for the help + fast response, truly appreciated!
